### DISCUSSION TITLE
Replicate osu!stable's gameplay animations more accurately

### DISCUSSION THREAD

**Discussion by refleraga** [ppy/osu#13660] [Ideas] [answered]
Currently **some animations are slightly different on lazer compared to stable**, such as:

- the circle hit animation being "faster"—it's not actually faster, but the fade out curve is different, making it look like it is
- slider follow circle animation at the end of sliders being different from stable

Please see the video linked [here](https://github.com/ppy/osu/discussions/13660#discussioncomment-923124) for a more detailed analysis of the above.

In addition, see [this](https://reflega.s-ul.eu/fg330w7X) video comparing both stable and lazer gameplay (best observed with fullscreen). I've removed anything that could potentially distract from the animations.

At first glance this looks to be negligible, but I believe (not with full certainty) that this contributes to how a lot of players say/complain that osu!lazer feels different to play (even with the Classic mod on) but aren't able to pinpoint on what exactly it is that's different. Getting lazer's hit animations as close to stable's might alleviate this.

**Comment by smoogipoo** [comment]
The slider follow circle one was intended at the time of implementation to keep things simple and not have a "virtual" size (this way, the input exactly matches the visuals). There's a few more subtle differences in that area, such as the circle should scale slightly when a tick is hit, with that animation being cut short as the next tick is approached.

But perhaps this is something that can be considered if there are any brave adventurers.

**Comment by peppy** [comment]
You mention "circle animations" but unless you are willing to provide frame comparisons, we are aiming to match these 1:1. All existing implementations have been done working with osu-stable code as a reference point.

**Reply by refleraga** [reply]
> unless you are willing to provide frame comparisons

[Here](https://reflega.s-ul.eu/bVLAUN8u). I hope this suffices.

What I've found:

- stable and lazer circles disappear at exactly the same time, but the fade out curve for lazer circles is different (steeper, maybe) than stable, making it look like they disappear faster
- lazer follow circle at the end of sliders fades out much faster and shrinks completely, compared to stable where it only shrinks down to a certain size

The follow circle animation at the start of sliders looks to be the same with stable, which I didn't notice before (my bad) so there's no problem there.

I'll edit my original post above with this new info.

**Reply by frenzibyte** [reply]
Thanks for the detailed video, that clears things up. 

Looking at the circle animation part, it seems that osu!stable (from the video) actually fades out in a linear (no easing) curve, which would look like this in osu!lazer:

https://user-images.githubusercontent.com/22781491/123496042-cf518d00-d62e-11eb-8d0f-8ccc52907ff4.mp4

But can't confirm any further without osu!stable code, so would like someone to double-check that before I proceed with the change.

**Reply by smoogipoo** [reply]
Also of note is that the follow circle size is different. I had another user point this out to me but hadn't taken the time to reproduce in this fashion yet and took it to be an issue with the skin.

**Reply by refleraga** [reply]
Indeed the follow circle size is a bit larger on lazer.

Since these are planned to be changed to match stable, should I create issues for these?

**Reply by peppy** [reply]
I'm 99% sure it's bigger to match that actual follow-able area. If anything, this would be implemented as a classic mod setting, I believe.

**Comment by frenzibyte** [comment, marked as answer]
Upon comparing with the [stable video](https://github.com/ppy/osu/discussions/13660#discussioncomment-923124) linked above, the slider follow circle animation has been updated to match stable (https://github.com/ppy/osu/pull/19090):

https://user-images.githubusercontent.com/22781491/196508910-3f342255-3132-4c81-9247-a4b93928f8eb.mp4

Hitobject fade easing is tracked under https://github.com/ppy/osu/issues/20806.
